docs/openfaas-pro/jetstream.md

```diff
@@ -18,17 +19,155 @@ On our blog we demo and explore some architectural patterns for these uses cases
+3. A Consumer is a group that various Subscribers can join to read messages and keep track of where they left off
+5. A Subscriber is what the queue-worker creates, if the max_inflight is set to 25, the queue-worker will create 25 subscribers
+
+> You can learn more about JetStream here: [Docs for JetStream](https://docs.nats.io/nats-concepts/jetstream)
+
+## Installation
+
+For staging and development environments OpenFaaS can be deployed with an embedded version of the NATS server which uses an in-memory store.
+
+To enable JetSteam for OpenFaaS set `jetstream` as the queue mode in the values.yaml file of the [OpenFaaS Helm chart](https://github.com/openfaas/faas-netes/blob/master/chart/openfaas/README.md)
+
+```yaml
+queueMode: jetstream
+nats:
+    streamReplication: 1
+```
+
+If the NATS pod restarts, you will lose all messages that it contains. In your development or staging environment, this shouldn't happen very often.
+
+For production environments you will need to install NATS separately using its Helm chart with at least 3 server replicas, so that if a pod crashes, the data can be recovered automatically.
+
+```yaml
+queueMode: jetstream
+nats:
+  streamReplication: 3
+  external:
+    enabled: true
+    host: "nats.nats"
+    port: "4222"
+```
+
+## Configure JetStream
+
+Every OpenFaaS async queue requires a Stream and Consumer to be created on the JetStream server. The queue-worker creates these on first startup if they do not exist. The default Stream and Consumer have some limits:
+
+- The maximum concurrency of each queue is limited to 512.
+- There is a limit on the amount of messages that can be queued for retries. The queue-worker will stop processing new messages when this limit is reached.
+
+To extend these limits Streams and Consumers can be created manually.
+
+### Configure Streams and Consumers manually
+
+Streams and Consumers can be defined manually, typically using the [NATS CLI tool](https://docs.nats.io/running-a-nats-service/configuration/resource_management/configuration_mgmt/nats-admin-cli). 
+
+>A [Kubernetes controller](https://docs.nats.io/running-a-nats-service/configuration/resource_management/configuration_mgmt/kubernetes_controller) is available for managing Streams and Consumers declaratively. This can be used if you are using a CD system like ArgoCD or Flux.
+
+You can use [arkade](https://github.com/alexellis/arkade) to install the NATS CLI.
+```
+arkade get nats
+```
+
+Port forward the nats server to your localhost so you can use the cli to interact with it.
+```bash
+kubectl port-forward -n openfaas svc/nats 4222:4222
+```
+#### Create a Stream
+
+The Stream will need to be created first. In this example we will create the Stream for the shared queue, `faas-request`.
+
+```bash
+nats stream create faas-request \
+  --replicas=1 \
+  --subjects=faas-request \
+  --storage=file \
+  --retention=limits \
+  --discard=old \
+  --max-msgs=-1 \
```

Comment:
Are any of these parameters defaults which we can omit for brevity?

Does a user really have to specify all of these flags just to extend the maximum number of messages for the consumer to have in flight?

Reply:
I reduced the list to the minimum required flags for our configuration. Users will be prompted for the remaining values and can use the defaults or their own values for the stream limits.